### DISCUSSION TITLE
How do I call a struct ARRAY data from SimpleStorage to StorageFactory?

### DISCUSSION THREAD

**Discussion by andersoncampolina** [smartcontractkit/full-blockchain-solidity-course-py#1567] [Q&A]
```
function sfGet(uint256 _simpleStorageIndex) public view returns (uint256) {
    return SimpleStorage(address(simpleStorageArray[_simpleStorageIndex])).retrieve();
}
```
on original course, he just do this... but, what if he wanna do the same thing with the People[] array? how do we call its data from another contract?

I added a call function on SimpleStorage, to call People[] data, and it works! it returns a tuple with the data, here the code:

```
function retrievePeople() public view returns (People[] memory) {
    return people;
}
```

But when I try to call it from StorageFactory, im trying this, but dont work:

```
function sfGetPeople(uint256 _simpleStorageIndex) public view returns (People[] memory) {
    return SimpleStorage(address(simpleStorageArray[_simpleStorageIndex])).retrievePeople();
}
```

I appreciate if someone could help :-)

**Comment by ghost** [comment]
I think:
- If you made `StorageFactory is SimpleStorage` then StorageFactory has access to `people` array
- I'm not sure if solidity support `returns (People[] memory)` but here is my alternative:
  - get `People` array length:
  ```js
      function getPeopleArrayLength() public constant returns(uint count) { return people.length; }
   ```
   - get a struct you want to access

**Reply by cromewar** [reply]
This is the best approach, does this helps @Zehn1984?

**Reply by andersoncampolina** [reply]
Im very glad to you to help me with this issue. Im gonna test it and reply the result. thanks anyway :-)

**Reply by andersoncampolina** [reply]
Didnt work, but I think is because I dont know how to do it. Let me post all the code here, so may more easy to you guys help.

First code:
```Solidity
// SPDX-License-Identifier: MIT

pragma solidity ^0.8.15;

contract CriaCarteirinha {
    // Struct é como uma variavel-objeto python é declarada no solidity
    // Criando estrutura do torcedor
    struct Torcedor {
        uint256 cpf;
        string nome;
        string endereco;
        uint256 telefone;
        string email;
    }

    // Criando estrtutura do historico
    struct Historico {
        string conquista;
    }

    // criando uma array do tipo objeto torcedor e historico.
    Torcedor[] public torcedor;
    Historico[] public historico;
    
    // faz o mapeamento de variaveis. Em solidity precisamos criar essa relaçao entre 2 variaveis para chamar um conteudo de uam variavel atravez de outra. nesse caso vai ser CPF => dado solicitado.
    mapping(uint256 => string) public cpfBuscaNome;
    mapping(uint256 => string) public cpfBuscaEndereco;
    mapping(uint256 => uint256) public cpfBuscaTelefone;
    mapping(uint256 => string) public cpfBuscaEmail;

    // adiciona torcedores à array Torcedor[]
    // em solidity existe 2 formas de armazenar os dados: memory - so armazena durante a execuçao da funcao / storage - fica armazenado em todo o programa
    function adicionarTorcedor(uint256 _cpf, string memory _nome, string memory _endereco, uint256 _telefone, string memory _email) public {
        torcedor.push(Torcedor(_cpf, _nome, _endereco, _telefone, _email));
        // estas proximas linhas so vão funcionar se a lista foi mapeada anteriormente.
        cpfBuscaNome[_cpf] = _nome;
        cpfBuscaEndereco[_cpf] = _endereco;
        cpfBuscaTelefone[_cpf] = _telefone;
        cpfBuscaEmail[_cpf] = _email;
    }

    // Adiciona conquistas à array historico.
    function adicionarConquistaHistorico(string memory _conquista) public {
        historico.push(Historico(_conquista));
    }

    function pegarDadosTorcedor() public view returns (Torcedor[] memory) {
        //retorna o resultado da função retrieve() do contrato 
        return torcedor;
    }

    function pegarDadosHistorico() public view returns (Historico[] memory) {
        //retorna o resultado da função retrieve() do contrato 
        return historico;
    }
}
```
Second Code (this one is who interacts with the first one:

// SPDX-License-Identifier: MIT

pragma solidity ^0.8.15;

//o objetivo do FabricaCarteirinha é fazer deploys e interações do contrato carteirinha.sol atravez de funções chamadas neste contrato.

import "./CriaCarteirinha.sol";

contract FabricaCarteirinha is CriaCarteirinha {
    
    CriaCarteirinha[] public carteirinhaArray;

    function criarCarteirinha() public {
        CriaCarteirinha carteirinha = new CriaCarteirinha();       
        carteirinhaArray.push(carteirinha); // ISSO RETORNA O ENDEREÇO HASH DO NFT / CARTEIRINHA CRIADO - ARMAZENAR EM BANCO DE DADOS SQL
    }

    function fcAdicionarTorcedor(uint256 _carteirinhaArrayIndex, uint256 _cpf, string memory _nome, string memory _endereco, uint256 _telefone, string memory _email) public {
        CriaCarteirinha(address(carteirinhaArray[_carteirinhaArrayIndex])).adicionarTorcedor(_cpf, _nome, _endereco, _telefone, _email);   
    }

    function fcAdicionarHistorico(uint256 _carteirinhaArrayIndex, string memory _conquista) public {
        CriaCarteirinha(address(carteirinhaArray[_carteirinhaArrayIndex])).adicionarConquistaHistorico(_conquista);   
    }

    function consultaDadosTorcedor(uint256 _carteirinhaArrayIndex) public view returns (CriaCarteirinha[] memory) {
        return CriaCarteirinha(address(carteirinhaArray[_carteirinhaArrayIndex])).pegarDadosTorcedor();
    }
}

**Reply by andersoncampolina** [reply]
on solidity things are kinda complicated. on JS I would just return carteirinha.pegasDadosTorcedor(); :-(
